python/pyarrow/array.pxi

```diff
@@ -1516,11 +1516,28 @@ cdef class Array(_PandasConvertible):
     def _to_pandas(self, options, types_mapper=None, **kwargs):
         return _array_like_to_pandas(self, options, types_mapper=types_mapper)
 
-    def __array__(self, dtype=None):
+    def __array__(self, dtype=None, copy=None):
+        if copy is False:
+            try:
+                values = self.to_numpy(zero_copy_only=True)
+            except ArrowInvalid:
+                raise ValueError(
+                    "Unable to avoid a copy while creating a numpy array as requested.\n"
+                    "If using `np.array(obj, copy=False)` replace it with "
+                    "`np.asarray(obj)` to allow a copy when needed"
+                )
+            # values is already a numpy array at this point, but calling np.array(..)
+            # again to handle the `dtype` keyword with a no-copy guarantee
+            return np.array(values, dtype=dtype, copy=False)
+
         values = self.to_numpy(zero_copy_only=False)
+        if copy is True and is_primitive(self.type.id) and self.null_count == 0:
+            # to_numpy did not yet make a copy
```

Comment:
Will this make a second copy for e.g. boolean?

Reply:
No, boolean is not considered as primitive (although you might expect that). So a boolean array takes the code path below just returning the numpy `values` is no dtype was required, so no second copy.

Reply:
No, sorry, boolean is of course "primitive", what I meant to say is that it is not "numeric", and the code above was meant to use `is_numeric` (I added that to libarrow.pxd for that purpose). Good catch ..